improve perfs

gurneyalex opened the pull request:
The choice of a suitable stock location was implemented using an iteration on all candidate locations  + several filters running for each. This is not efficient when we have several 1000s of locations (each
check was taking 30ms -> several minutes required to compute a putaway location). This commit reimplements the algorithm by using queries to select locations matching the rules. In order to be able to compute this using the ORM, we added some technical fields which should not harm performance too much.

gurneyalex committed:
Revert "Remove views too slow to show with hundreds of locations"

This reverts commit 2e516652452d97d1f9abcefc4db298b31ca10e8c.

gurneyalex committed:
improve perfs

The choice of a suitable stock location was implemented using an
iteration on all candidate locations + several filters running for each.
This is not efficient when we have several 1000s of locations (each
check was taking 30ms -> several minutes required to compute a putaway
location). This commit reimplements the algorithm by using queries to
select locations matching the rules. In order to be able to compute this
using the ORM, we added some technical fields which should not harm
performance too much.

gurneyalex commented:
This is best reviewed in "split" mode rather than "unified"

guewen commented on stock_storage_type/models/stock_location.py:
```diff
@@ -55,6 +54,81 @@ class StockLocation(models.Model):
+    def _compute_children_in_move_line_ids(self):
+        for rec in self:
+            rec.children_in_move_line_ids = (
+                rec.mapped('child_ids.in_move_line_ids') | rec.in_move_line_ids
```
It will contain only the lines of the current location and the direct children, not more than one level, is it what we expect?

gurneyalex replied:
nope, you are correct this was a late change when I noticed the search on stock.move.lines was different from the one on stock.move

gurneyalex replied:
should be fixed now

guewen commented on stock_storage_type/models/stock_location.py:
```diff
@@ -153,125 +233,98 @@ def get_storage_locations(self, products=None):
-        domain = self._prepare_existing_domain(base_domain, products=products)
-        return self.env["stock.move"].search(domain, limit=1)
+        pertinent_loc_storage_types = StockLocationStorageType.search(
+            pertinent_loc_storagetype_domain
```
The building of `pertinent_loc_storagetype_domain` should be extracted in a method to allow customizations.

gurneyalex replied:
done

guewen commented on stock_storage_type/models/stock_location.py:
```diff
@@ -153,125 +233,98 @@ def get_storage_locations(self, products=None):
+                    location_domain.append(
+                        ('location_will_contain_lot_ids', 'not in', lots.ids),
+                    )
+            locations = self.search(location_domain, limit=limit)
```
Domain to extract as well

gurneyalex replied:
done

guewen reviewed (commented):
That's great, some improvements regarding extensibility can be done. But the only dubious thing to me is the function field looking at one level of children.

@grindtildeath  may have comments regarding integration of chaotic ABC

gurneyalex committed:
fixup! improve perfs

grindtildeath commented on stock_storage_type/models/stock_location.py:
```diff
@@ -55,6 +54,81 @@ class StockLocation(models.Model):
+                | rec.mapped("children_in_move_line_ids.lot_id")
+            )
+
+    @api.depends('quant_ids', 'in_move_ids', 'in_move_line_ids')
```
```suggestion
    @api.depends('quant_ids', 'in_move_ids', 'children_in_move_line_ids')
```

grindtildeath commented on stock_storage_type/models/stock_location.py:
```diff
@@ -55,6 +54,81 @@ class StockLocation(models.Model):
+        "the location, either now or in pending operations"
+    )
+
+    @api.depends('child_ids.children_in_move_line_ids')
```
```suggestion
    @api.depends('in_move_line_ids', 'child_ids.children_in_move_line_ids')
```

gurneyalex replied:
no, I'm not using in_move_line_ids in the method.

guewen replied:
actually you are :)

grindtildeath commented on stock_storage_type/models/stock_location.py:
```diff
@@ -98,45 +172,51 @@ def _get_pack_putaway_strategy(self, putaway_location, quant, product):
-            pref_loc = pack_loc.location_id
+        if not package_locations:
+            return dest_location
+        for pref_loc in package_locations.mapped('location_id'):
```
Are we sure to keep the same order of `package_locations` while using `mapped` here? :thinking:

guewen replied:
`mapped()` uses [union](https://github.com/odoo/odoo/blob/13.0/odoo/models.py#L5535-L5536) which keep the "first occurrence order" so it looks correct

grindtildeath commented on stock_storage_type/models/stock_location.py:
```diff
@@ -98,45 +172,51 @@ def _get_pack_putaway_strategy(self, putaway_location, quant, product):
             if (
                 pref_loc.pack_putaway_strategy == "none"
-                and pref_loc._package_storage_type_allowed(
+                and pref_loc.select_allowed_locations(
```
Here, we should only check if `package_storage_type` is allowed into `pre_loc` (according to its stock.location.storage.type). So I feel like all the searches on `stock.location` done in `select_allowed_locations` could be spared...

grindtildeath commented on stock_storage_type/models/stock_location.py:
```diff
@@ -153,125 +233,96 @@ def get_storage_locations(self, products=None):
-            ):
-                allowed_ids.add(location.id)
-        return self.browse(allowed_ids)
+    def _domain_location_storage_type_constraints(self, compatible_locations, package_storage_type, quants, products):
```
```suggestion
    def _domain_location_storage_type_constraints(self, package_storage_type, quants, products):
```

grindtildeath commented on stock_storage_type/models/stock_location.py:
```diff
@@ -153,125 +233,96 @@ def get_storage_locations(self, products=None):
+        # we consider.
+        StockLocationStorageType = self.env['stock.location.storage.type']
+        compatible_location_storage_types = StockLocationStorageType.search(
+            [('allowed_location_ids', 'in', compatible_locations.ids)]
```
```suggestion
            [('allowed_location_ids', 'in', self.ids)]
```

grindtildeath commented on stock_storage_type/models/stock_location.py:
```diff
@@ -153,125 +233,96 @@ def get_storage_locations(self, products=None):
-            not self.allowed_location_storage_type_ids or allowed_location_storage_types
+        pertinent_loc_s_t_domain = (
+            LocStorageType._domain_location_storage_type_constraints(
+                compatible_locations, package_storage_type, quants, products
```
```suggestion
            compatible_locations._domain_location_storage_type_constraints(
                package_storage_type, quants, products
```

grindtildeath commented on stock_storage_type/models/stock_location_storage_type.py:
```diff
@@ -114,3 +118,61 @@ def _compute_has_restrictions(self):
+        return location_domain
+
+    @api.model
+    def _domain_location_storage_type_constraints(self, compatible_locations, package_storage_type, quants, products):
```
This one is already defined in `stock_location.py`. But to me it makes more sense to keep it there since `compatible_locations` must be a `stock.location` and can be passed on `self`.

guewen replied:
~Pushed a fixup for this~ no permission

grindtildeath commented on stock_storage_type/models/stock_location.py:
```diff
@@ -4,9 +4,8 @@
 
 from odoo import _, api, fields, models
 from odoo.exceptions import ValidationError
-from odoo.osv.expression import AND, OR
```
IMO, let's try to keep these operators when building complex domain for the sake of understandability :wink:

guewen commented:
8 tests are failing :( I planned to take over for the few changes but not sure I'll have the time to find the cause of the failures.

Commit:
fixup! fixup! improve perfs

guewen commented on stock_storage_type/models/stock_location.py:
```diff
@@ -55,6 +54,81 @@ class StockLocation(models.Model):
+
+    in_move_ids = fields.One2many(
+        'stock.move', 'location_dest_id',
+        domain=[('state', 'not in', ('draft', 'done', 'cancel'))],
```
Shouldn't we use IN with the other states to be sure the index is used?

gurneyalex replied:
It could help, but I'm not 100% sure the index will be used.

guewen commented on stock_storage_type/models/stock_location_storage_type.py:
```diff
@@ -114,3 +118,61 @@ def _compute_has_restrictions(self):
+            )
+        if self.do_not_mix_products:
+            location_domain.append(
+                ('location_will_contain_product_ids', 'not in', products.ids)
```
This condition prevent to put a good in a location if it contains the same product, which is not what we want,
we want to exclude a location if it contains other products. (same thing for the lot below)

Commit:
fixup! fixup! fixup! improve perfs

jbaudoux commented on stock_storage_type/models/stock_location.py:
```diff
@@ -153,125 +234,95 @@ def get_storage_locations(self, products=None):
+            pertinent_loc_storagetype_domain += [
+                '|',
+                ('max_height', '=', 0),
+                ('max_height', '>', quants.package_id.height),
```
```suggestion
                ('max_height', '>=', quants.package_id.height),
```

jbaudoux commented on stock_storage_type/models/stock_location.py:
```diff
@@ -153,125 +234,95 @@ def get_storage_locations(self, products=None):
+            pertinent_loc_storagetype_domain += [
+                '|',
+                ('max_weight', '=', 0),
+                ('max_weight', '>', quants.package_id.pack_weight),
```
```suggestion
                ('max_weight', '>=', quants.package_id.pack_weight),
```

jbaudoux reviewed (changes requested):
Small remark regarding max_height/weight

Commit:
fixup! fixup! fixup! fixup! improve perfs

Commit:
fixup! fixup! fixup! fixup! fixup! improve perfs

jbaudoux commented on stock_storage_type/models/stock_location.py:
```diff
@@ -55,6 +54,86 @@ class StockLocation(models.Model):
+    def _compute_location_is_empty(self):
+        for rec in self:
+            if (
+                    rec.quant_ids
```
There could be a quant with quantity = 0. In this case the location is still empty

jbaudoux commented on stock_storage_type/models/stock_location_storage_type.py:
```diff
@@ -114,3 +111,39 @@ def _compute_has_restrictions(self):
+            ('id', 'in', candidate_locations.ids),
+            ('allowed_location_storage_type_ids', '=', self.id),
+        ]
+        if self.only_empty:
```
From here, it is not related to storage type. Any chance to extract this in another method that is not recomputed for each storage type?

guewen replied:
It's related to the configuration of the storage type (only_empty, do_not_mix_products, do_not_mix_lots).
We could optimize this by grouping the storage types by similar configuration to reduce the number of queries, but it isn't required now. Adding a todo

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! improve perfs

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! improve perfs

guewen commented:
@gurneyalex @grindtildeath in my last commit https://github.com/grindtildeath/wms/pull/6/commits/ee8e96d256be98211e9477060f750e9eb8737376, I could make more tests to pass, but I had to new dependencies on compute methods and to add `rec.mapped("in_move_line_ids.product_id")` in the computation of `_compute_location_will_contain_product_ids` and the same for lots, although, `in_move_line_ids` is normally comprised in `children_in_move_line_ids` already! Not very happy with this but I'm not sure I can it differently now.

Yet 3 tests to make pass.

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! improve perfs

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! improve perfs

jbaudoux commented on stock_storage_type/models/stock_location.py:
```diff
@@ -153,125 +251,109 @@ def get_storage_locations(self, products=None):
-                extra_domain = OR([extra_domain, [("lot_id", "!=", lot.id)]])
-            domain = AND([domain, extra_domain])
-        return domain
+    def _get_ordered_children_locations(self):
```
I think we want only leaf locations and not the intermediate nodes in the location structure. Same for ABC

guewen replied:
Yes! That's exactly what I was discussing with @grindtildeath at the moment and started to change

guewen commented:
The field `children_in_move_line_ids` is never recomputed, in the tests at least

guewen commented:
> The field `children_in_move_line_ids` is never recomputed, in the tests at least

But do we actually need to include the children location in the domains? A put-away is meant to be in a leaf, no?

jbaudoux commented on stock_storage_type/models/stock_location.py:
```diff
@@ -55,6 +54,98 @@ class StockLocation(models.Model):
+                sum(rec.quant_ids.mapped("quantity"))
+                or rec.in_move_ids
+                or rec.in_move_line_ids
+                or rec.children_in_move_line_ids
```
This already includes the above in_move_line_ids

guewen replied:
Yes, but I removed the children_in_move_line_ids field in the latest fixups (see my last comments), because:

* children_in_move_line_ids was not recomputed, not sure why but didn't investigate further as:
* I don't see any scenario where we are interested in what moves lines goes into the child locations are we are interested only by the leaf locations

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! improve perfs

Commit:
Find leaf sub-locations instead of all children

We will never do a put-away in an intermediate location,
so they should never be kept as candidate locations for the
put-away.

guewen commented:
Merged in https://github.com/OCA/wms/pull/12
